threading branch

vtjnash opened the pull request:
this closes https://github.com/JuliaLang/julia/issues/10772 and superceeds the other threading branches

the goal of this is to have the threading support off by default (controlled by a flag in julia.h), but to have the code building and passing tests either way; and to get this merged into master sooner rather than later.

kpamnany committed:
Merging threads branch.

kpamnany committed:
Fix for the parallel for form.

Commit:
Change LLVM_GIT_URL_LLVM to use "https:" instead of "git@", since some
firewalls cannot deal with the latter.

JeffBezanson committed:
inlining improvement that helps #10527

this avoids inserting temporary variable copies of local variables
unless they might be assigned by an inner function.

Conflicts:
	base/inference.jl

JeffBezanson committed:
fix merge mistake in fix for #10527

Commit:
Cosmetic change to bring closer to master: __JL_THREAD -> JL_THREAD

Commit:
More cosmetic changes to bring closer to master.

Commit:
Bring tasking up to version in master branch.

Commit:
Merge branch 'adr/snapshot' of https://github.com/JuliaLang/julia into adr/threading,
which in turn is a snapshot of threading branch.

Conflicts:
	base/cartesian.jl
	base/inference.jl
	src/Makefile
	src/codegen.cpp
	src/gc.c
	src/init.c
	src/julia.h
	src/task.c
	test/runtests.jl

Commit:
Fix Relocation and CodeModel.

Commit:
Merge branch 'adr/snapshot' of https://github.com/JuliaLang/julia into adr/threading

Commit:
Use https://github.com/JuliaLang/llvm.git -b kf/tlsrebase5

kpamnany committed:
Affinity fix for threads. Multi-threaded workloads.

Commit:
Resolve merge of base/inference.jl

Keno committed:
Some whitespace fixes

kpamnany committed:
Lattice-Boltzmann 3D workload.

kpamnany committed:
Minor updates & fixes to multithreaded workloads.

ViralBShah committed:
Seed MT differently on each proc

vtjnash committed:
Merge branch 'adr/threading' of https://github.com/JuliaLang/julia into master

vtjnash committed:
remove compiler overrides

vtjnash committed:
for TLS variables, need to call dlsym rather than & to get the address of the metadata rather than the thread-local pointer

vtjnash committed:
reduce cruft from codegen init

vtjnash committed:
makefile fixes to build on darwin

JeffBezanson committed:
put GC barrier back

vtjnash committed:
remove accidental double-init of jlexc_var

vtjnash committed:
fix gc-verify code to work with threading

vtjnash committed:
more gc fixes and cleanup

JeffBezanson committed:
some 0.4 updates to threads code; get test-threads passing again

vtjnash committed:
fix another translation error in the gc-verify code

vtjnash committed:
syncronize contentious gc state

JeffBezanson committed:
synchronize type caches

rename gc lock to pagealloc, since that's all it actually protects

JeffBezanson committed:
synchronize add_finalizer. fixes #10718

vtjnash committed:
add JULIA_ENABLE_THREADING guard block around all threading-specific code

vtjnash committed:
fixup for jn/threading build

JeffBezanson committed:
slightly reduce overhead in jl_threading_run

JeffBezanson committed:
fix a `return` that needs to be `continue` with the function body wrapped in FOR_EACH_HEAP

JeffBezanson committed:
fix build of gc.c when threads are disabled

JeffBezanson committed:
mark all thread stacks in GC

JeffBezanson committed:
hoist loads of constant closed variables

improves the code generated for closures. slightly helps #10752

Commit:
Atomics.

Commit:
@nany exported; was already defined

Commit:
Reduce default number of threads to 4.

Commit:
Refactor threading stuff into Base.Threads

The Threads module now includes the threading constructs (@threads),
the thread management functions (threadid(), maxthreads(), nthreads()),
atomics (Atomic, atomic_cas!, atomic_xchg!, etc.), and locks (SpinLock,
RecursiveSpinLock, Mutex, lock! and unlock!).

Some test cases are in test/threads.jl. Performance tests are updated
in test/perf/threads/.

vtjnash committed:
fix whitespace script to not get confused by glob expansion, and whitespace errors

vtjnash committed:
Merge remote-tracking branch 'origin/master' into jn/threading

Conflicts:
	Make.inc
	contrib/README.md
	src/Makefile
	src/codegen.cpp
	src/gc.c
	src/init.c

vtjnash committed:
set a min target architecture of pentium4 to require support for sse2 and _mm_pause

vtjnash committed:
Uint => UInt deprecation

vtjnash committed:
get threads test passing without threading support

vtjnash committed:
add support for llvm < v3.6 to atomics

timholy commented:
Wow, really exciting stuff!

vtjnash committed:
correct size of uv_mutex on windows (& everywhere)

johnmyleswhite commented:
It's like Christmas day over here

tkelman commented:
Everything with an Intel license needs to be spelled out in `LICENSE.md` and `contrib/add_license_to_files.jl`.

tkelman commented on Make.inc:
```diff
@@ -496,6 +494,8 @@ else
 XC_HOST := $(ARCH)$(shell echo $(XC_HOST) | sed "s/[^-]*\(.*\)$$/\1/")
 MARCH := $(subst _,-,$(ARCH))
+else # insert ARCH into HOST
+XC_HOST := $(ARCH)$(shell echo $(BUILD_MACHINE) | sed "s/[^-]*\(.*\)$$/\1/")
```
what's the purpose of always setting `XC_HOST` even when not cross-compiling?

vtjnash replied:
if you explicitly set `ARCH`, then it assumes you meant to try to do a cross-compile. without this, libunwind was getting confused by the mismatch between `--host` and `CC=gcc -m32`.

tkelman replied:
Good to know, thanks. Wouldn't that have happened independently of the threading branch though?

vtjnash replied:
sure. i don't go around randomly testing build configurations however, so i only notice these issues when building branches.

tkelman commented on test/perf/threads/lbm3d/lbm3d.jl:
```diff
@@ -0,0 +1,215 @@
+
+
+function relax!(F, UX, UY, UZ, nx, ny, nz, deltaU, t1D, t2D, t3D, sSQU)
+  tid = threadid()
```
4-space this file

Keno commented:
If they're gonna be part of the base library, they should have our standard license header.

Keno commented:
@kpamnany Might be good to just run that by Intel Legal. Shouldn't be a problem. Most other open source projects have the same policy (e.g. LLVM which Intel contributes to heavily).

Commit:
Style: 4-spaced.

kpamnany commented:
Have inquired and will get back on changing the license to MIT.

kpamnany commented:
Confusing build errors coming from code under LLVM38. Calling ``createTypeBasedAAWrapperPass`` but it doesn't exist in ``deps/llvm-svn``. How was this compiling previously?

StefanKarpinski commented:
llvm-svn is a moving target?

kpamnany commented:
This branch of ``llvm-svn`` shouldn't be moving (elegantly named ``tlsrebaseagainagain``).

Keno commented:
This PR is rebased against master. Does it still include that?

kpamnany commented:
Yes.
```
~/13410/deps/srccache/llvm-svn$ git branch
* jn/tlsrebaseagainagain
```

Keno commented:
Ok, then it's probably the other way around. Master is head in llvm version of threading, so we might need to rebase again.

vtjnash commented:
In order to make this merge-ready with master, I removed the Make.user customizations, including the llvm branch specification.

kpamnany commented:
Right. It builds fine with ``JL_ENABLE_THREADING`` turned off, and with whatever is LLVM default.

But when you turn on ``JL_ENABLE_THREADING`` and add in llvm-svn into Make.user as we need it for threading, the build fails with multiple errors in ``codegen.cpp``, all under ``#ifdef LLVM38``.

Keno commented:
I think you might just have a stale branch in deps/srccache, which is causing your build failures.

kpamnany commented:
This is a fresh clone, to test exactly this.

vtjnash commented:
the primary blocker for me on this PR right now is that gcc4/5 doesn't compile Julia correctly with `-m32 -march=pentium4`, so the SSE2 instruction `pause` cannot be used. In particular, in the expression in `jl_compute_field_offsets`:
```c
            size_t alsz = (((sz) + (al-1)) & (-al));
            if (alsz > sz)
                st->haspadding = 1;
```
in the assembly output, it has effectively turned that conditional into `if (sz > sz)`

I don't think I see any undefined behavior here according to the C standard, so I filed a bugreport against gcc: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=67814

tkelman commented:
Any hope of a workaround? GCC bug is not a good time scale on this.

Keno commented:
inline assembly?

kpamnany commented:
I still see those compile errors in ``codegen.cpp``. What's wrong there?

vtjnash committed:
work around a gcc register allocation error

see https://gcc.gnu.org/bugzilla/show_bug.cgi?id=67814

vtjnash committed:
fix minor makefile grievances

vtjnash committed:
Merge remote-tracking branch 'origin/master' into jn/threading

Conflicts:
	Makefile
	contrib/windows/msys_build.sh

vtjnash commented:
@kpamnany can you change the license on the files to the standard text ("# This file is a part of Julia. License is MIT: http://julialang.org/license\n\n")? then I think this should be ready to merge.

Commit:
Updated license.

kpamnany commented:
Done. But I still see those compile errors with ``JULIA_ENABLE_THREADING`` turned on.

Keno commented:
Just needs a rebase of the threading patches on top of llvm svn.

Keno commented:
I'd the locking implementation to be available even if threads a are disabled to separate the profiler from trying to unwind while we're modifying libunwind's internal data structures.

Keno commented:
But that can come after merging.

tkelman commented on base/locks.jl:
```diff
@@ -0,0 +1,138 @@
+
+export SpinLock, Mutex, init_lock!, destroy_lock!, lock!, trylock!, unlock!
+
+abstract Lock
```
should this be called `AbstractLock` ?

Keno committed:
Fix atomics.jl on llvm-svn

"svn" is not a `VersionNumber`

kpamnany commented:
Expect lots of input and suggestions on desired interface and functionality. These can go on top of the list for after merge. :)

tkelman commented:
What are the licenses on the original code for `stockcorr_vec.m` and `lbm3d.m` ? I guess we already had the stockcorr benchmarks at https://github.com/JuliaLang/julia/blob/master/test/perf/kernel/stockcorr.m, but did anyone ever ask the original author for permission?

IainNZ commented on base/cartesian.jl:
```diff
@@ -98,6 +98,19 @@ function _nall(N::Int, criterion::Expr)
+    _nany(N, criterion)
+end
+
+function _nany(N::Int, criterion::Expr)
```
Tests?

kpamnany replied:
I'd add a couple, but I can't find tests for any of the existing macros in ``Cartesian``.

amitmurthy commented on base/locks.jl:
```diff
@@ -0,0 +1,138 @@
+# This file is a part of Julia. License is MIT: http://julialang.org/license
```
There is already a `base/lock.jl`. Shouldn't the two be merged?

kpamnany replied:
Yes. I'm working on the interaction between threads and tasks, which is also needed for I/O. When that is properly designed, these will get merged.

tkelman commented:
Would need the following patch to build under msvc: https://gist.github.com/676fffb916a7d67f4a59

Anyone opposed to committing that here?

vtjnash commented on src/options.h:
```diff
@@ -17,11 +17,8 @@
-
 // (Experimental) Use MCJIT ELF, even where it's not the native format
-// #define FORCE_ELF
+#define FORCE_ELF
```
this line might need to be reverted before merging. @Keno

Commit:
Added credits.

kpamnany commented:
@tkelman: I emailed the author and he has updated the post at http://www.walkingrandomly.com/?p=3604 to state that the code is licensed under the 3 clause BSD license. He's also asked for his name and a link as credit; I've added those.

Commit:
Change Lock to AbstractLock and fix compile error

tkelman commented:
@kpamnany great, that is good news, thanks for following up on that (and to the author for responding quickly). I think we do need to put the license text in accompanying the source code, either inline or in the same directory.

ViralBShah commented:
@kpamnany Thanks. I think the code had come up on the mailing list, and I had put it in our benchmarks. @tkelman Thanks for noticing.

tkelman committed:
Fixes for MSVC build

vtjnash committed:
Merge remote-tracking branch 'origin/master' into jn/threading

vtjnash committed:
use llvm-config directly to always determine the real llvm-version

this reverts f5da3f8aba86d14e32577ea9dc6e7f0993c2a748

vtjnash commented:
please use the llvm-svn branch `jn/tls37` now. i'm getting ready to merge this, so let me know if there remaining problems.

tkelman commented:
`lbm3d.m` is still missing a license

kpamnany commented:
@vtjnash: with threading turned on, there are no compile errors with ``jn/tls37``, and threads seem to work fine.

@tkelman: I've emailed the author of LBM. Awaiting a response.

tkelman commented on src/support/dtypes.h:
```diff
@@ -137,6 +137,14 @@
 #  define NOINLINE_DECL(f) f __attribute__((noinline))
 #endif
 
+#if defined(_OS_WINDOWS_) && !defined(_COMPILER_MINGW_)
```
why is this duplicated?

tkelman replied:
any reason here, or was this just a wrong merge resolution?

vtjnash replied:
looks like a merge tool error

iamed2 commented:
I'm getting compile errors with threading turned on:
```
    CC src/codegen.o
/Users/ericdavies/repos/juliathreading/src/codegen.cpp:5350:54: error: no member named 'GeneralDynamicTLSModel' in 'llvm::GlobalValue'
                           NULL, "jl_pgcstack", NULL JL_THREAD_MODEL);
                                                     ^~~~~~~~~~~~~~~
/Users/ericdavies/repos/juliathreading/src/codegen.cpp:5343:39: note: expanded from macro 'JL_THREAD_MODEL'
#define JL_THREAD_MODEL ,GlobalValue::GeneralDynamicTLSModel
                         ~~~~~~~~~~~~~^
/Users/ericdavies/repos/juliathreading/src/codegen.cpp:5356:66: error: no member named 'GeneralDynamicTLSModel' in 'llvm::GlobalValue'
                           NULL, "jl_exception_in_transit", NULL JL_THREAD_MODEL);
                                                                 ^~~~~~~~~~~~~~~
/Users/ericdavies/repos/juliathreading/src/codegen.cpp:5343:39: note: expanded from macro 'JL_THREAD_MODEL'
#define JL_THREAD_MODEL ,GlobalValue::GeneralDynamicTLSModel
                         ~~~~~~~~~~~~~^
```
I'm on the `jn/tls37` branch and such.

kpamnany commented:
Threading requires a patched LLVM. Create a ``Make.user`` containing:

```
LLVM_VER = svn
LLVM_GIT_URL_LLVM = https://github.com/JuliaLang/llvm.git -b jn/tls37
LLVM_DEBUG = 0
LLVM_ASSERTIONS = 1
```

ranjanan commented:
@vtjnash Unfortunately, I keep getting this error:
```
LLVM ERROR: Cannot select: 0x1d7afd0: i64 = X86ISD::WrapperRIP 0x1d7aea8 [ORD=10] [ID=72]boot.jl:290
  0x1d7aea8: i64 = TargetGlobalTLSAddress<%jl_value_t*** @jl_pgcstack> 0 [TF=10] [ORD=10] [ID=42]boot.jl:290
In function: julia__new_0
make[1]: *** [/home/ranjan/julia-threading/usr/lib/julia/inference0.ji] Error 1
make: *** [julia-inference] Error 2
```
@kpamnany which platform did you compile on?

iamed2 commented:
@kpamnany Yes I have that and I get the errors I posted above.

kpamnany commented:
@iamed2: See https://github.com/JuliaLang/llvm/blob/jn/tls37/include/llvm/IR/GlobalValue.h#L118. Not sure why your build isn't seeing this definition. Do you have an old ``deps/llvm-svn`` hanging around?

@ranjanan, @vtjnash: confirming that LLVM error when trying to build on Linux. Builds on OS X though.

vtjnash commented:
Linux probably needs us to explicitly request PIC_ (https://github.com/JuliaLang/julia/pull/13410/files#diff-6d4d21428a67320600faf5a1a9f3a16aR6005). It's already the default on the other platforms, since it is invalid to build a shared library without it (as linux is probably trying to do here). I had issues with non-x86_64 builds with that configuration though, so this may be something in investigate further after merging the PR.

kpamnany commented:
Yeah, changing ``Reloc::Default`` to ``Reloc::PIC_`` at ``codegen.cpp:6005`` fixes this. I suggest making this the default as Linux x86_64 is probably the key platform for all this.

vtjnash commented:
that would potentially break the normal build on non-X86_64 linux systems, so it'll have to be done with some care

JeffBezanson commented:
I propose using an `#ifdef` :P

kpamnany commented:
We need to make it easier for people to try this out. Right now, to enable threading you have to uncomment [JULIA_ENABLE_THREADING](https://github.com/JuliaLang/julia/blob/jn/threading/src/julia.h#L70), and create a ``Make.user`` with ``LLVM_VER`` and ``LLVM_GIT_URL_LLVM`` set correctly, as described above. And for Linux x86_64, edit ``codegen.cpp``, also as described above.

Can we instead ask for a ``Make.user`` with something like ``JULIA_THREADING = 1`` in it? Can the remaining requirements be handled automatically? I'm not sure how ``LLVM_VER`` etc. would be updated -- ``Make.user`` is parsed only after ``Make.inc``, correct?

JeffBezanson commented:
+1 to that. Make.user is loaded early within Make.inc, so this should be possible.

iamed2 commented:
@kpamnany Nope, I do have that line in my GlobalValue.h. Fresh clone of julia, all the required make options (+ OPENBLAS_TARGET_ARCH=HASWELL). I verified that the correct version of llvm-svn is cloned.

kpamnany commented:
@iamed2: if you have the correct ``llvm-svn`` then sure, you must have the right ``GlobalValue.h`` in there but the error you're seeing suggests that ``codegen.cpp`` is being compiled with the wrong ``GlobalValue.h``; where is that one? Perhaps in ``$(JULIA_HOME)/usr/include/llvm/IR``? You could try ``make -n`` and look for line in which ``codegen.cpp`` is being compiled to find out.

ViralBShah commented:
Can we have a Make.user in this branch so that people can try the threading branch easily? Would be great to get it in master with `ifdef` and some basic travis testing.

JeffBezanson commented:
Let's go with @kpamnany 's idea, and implement a single Make.user flag to enable all of this, then we can just merge this branch into master.

iamed2 commented:
@kpamnany Ah yeah looks like it's pulling in the headers from my Homebrew llvm33-julia. `llvm-config --includedir` returns the correct directory though, so I don't know what's up with that. Seems like a problem in the build process somewhere?

tkelman commented:
Building the right version of llvm from scratch isn't going to work on travis.

> He says "I don't really have a license, and consider it free to use. If it's going on your website, I'd appreciate a link back to the source (i.e. the link you included in your original email), but that would be a courtesy rather than a requirement." This link is already in lbm3d.m so I think we're good with this.

Ugh. Putting it in our repository is saying we have copyright on the code to distribute it under the MIT license. If we really need to have the matlab code checked in here, we should add in a note indicating we have the author's permission to do this.

vtjnash commented:
> Ah yeah looks like it's pulling in the headers from my Homebrew llvm33-julia

don't install homebrew in /usr/local: problem solved.

iamed2 commented:
@vtjnash why does the Homebrew install location matter?

vtjnash commented:
/usr/local is always in the compiler search path, which makes it hard to build applications on that machine that don't have unexpected conflicts or dependencies inherited from homebrew

iamed2 commented:
llvm-config should handle this. The llvm-config built by Julia's build returns the correct locations.

I guess I should make a separate issue for this.

iamed2 commented:
Well, I `brew rm llvm33-julia`d and yet the build (from a fresh clone) is still trying to access it! I'm getting this as part of the build process:

```
make[1]: /usr/local/opt/llvm33-julia/bin/llvm-config-3.3: Command not found
make[1]: /usr/local/opt/llvm33-julia/bin/llvm-config-3.3: Command not found
make[1]: /usr/local/opt/llvm33-julia/bin/llvm-config-3.3: Command not found
make[1]: /usr/local/opt/llvm33-julia/bin/llvm-config-3.3: Command not found
```

Commit:
Licensing: added note for LBM3D.

Commit:
Threads: make it easier to turn on threads

Set JULIA_THREADS = 1 in Make.user to build with threading. The
necessary flags are conditionally set in Make.inc.

kpamnany commented:
@tkelman: added a note in ``lbm3d.m`` stating that we have the author's permission to reproduce the code. Would you go ahead and fix whatever's wrong in ``src/support/dtypes.h``?

@vtjnash and @JeffBezanson: if I did it right (tested on OS X), a ``Make.user`` containing ``JULIA_THREADS = 1`` should be all that is needed to enable threads.

Let's merge this.

hayd commented:
@kpamnany Perhaps it's worth asking explicitly if the author would willing to allow us to relicense this code under MIT. Their terms appear to be the same as MIT and it would be easier for julia (one less exception).

kpamnany commented:
@hayd, @tkelman: Author's response is "Hi Kiran. MIT licence agreed. I'm sure this email is enough, but I'll update the page also when I get round to it. Cheers, Iain."

tkelman committed:
remove duplicated NOINLINE defines

Commit:
Threads: add enable threading flag correctly

Commit:
Threads: use the correct flags and conditional

ViralBShah commented:
Bump. Merge this soon?

tkelman commented:
There's a conflict that needs resolving. I dislike merge commits since they can bury bad conflict resolutions along with thousands of lines of unreviewable diff of everything else that has happened on master, but this is too messy to rebase at this point. A comment about where conflicts were, or a manual commit that only touches the files where conflicts happened would help in tracking conflict resolutions.

vtjnash commented:
I'm not sure why github doesn't present the merge commits better. `git show` gives me only the lines that were affected by the merge conflict. in the future, I may try to use `git rebase --preserve-merges` to reduce the number of merge commits.

kpamnany commented:
Yay!

ihnorton commented:
:joy_cat:

johnmyleswhite commented:
:tada:

tkelman commented:
@vtjnash that is a great tip, thanks for that - https://gist.github.com/7e083009a110666331ba

vtjnash commented:
and even that is a massive over-approximation of the merge conflict (which was just resolving the order of the the assert vs JL_LOCK changes at the top of jl_trampoline_compile_linfo)

KristofferC commented:
:100:

JeffBezanson commented:
:fireworks:

ViralBShah commented:
:+1:

tkelman commented:
Here's a source build with all the deps in place for this: https://hub.docker.com/r/tkelman/julia-threads/

```
sudo docker run -t -i tkelman/julia-threads /bin/bash
cd /home/julia-x86_64
make -j4 # just julia src and sysimg, all deps are already built
```

KristofferC commented:
@tkelman Hmm, I seem to get stuck on second round of `inference.jl`..

tkelman commented:
odd, are you in a VM? maybe adding `MARCH=x86-64` might help?

KristofferC commented:
Yes, I used the "Kitematic"-thing for Docker on windows which seems to use VirtualBox. Seems to work now with `MARCH`

tkelman commented:
Probably a virtualbox bug with avx.

StefanKarpinski commented:
Getting this merge is so awesome.

KristofferC commented:
Since I am eager to play with this, is there any documentation yet for the new thread-macros or is it best to just look at the examples and go from there?

ViralBShah commented:
Best to look at the threads perf tests. The interface is very rudimentary. I don't think we should document until APIs stabilise.

kpamnany commented:
There are a few small example programs in ``test/perf/threads/``. The simplest use cases are in ``test/threads.jl`` which is kind of a tutorial.

The interface is, in fact, rudimentary and the syntax preliminary. However, I think the basic interface (``jl_threading_run(threadfun, args)``) is extremely powerful. Not only does it allow high performance parallel for loops (which might be the single most widely used parallelism construct in technical computing applications designed for scale), but together with atomics, which are also here, it can easily be used to build other shared memory parallelism models, from transactional memory to work-stealing. So, while there are certainly many many things remaining to do, I'm hoping many people will use this and offer feedback to drive priorities, or even better, write some code and raise some PRs.

@ViralBShah, @StefanKarpinski: what's a good venue to discuss the threading interface and improvements to it? Here? In https://github.com/JuliaLang/julia/issues/1790?

ViralBShah commented:
I think we should close #1790 and open specific issues to discuss various threading related issues and improvements.

KristofferC commented:
Using this branch I could speed up my kd tree creation with a factor of about 2.25x encountering very little problem. Really nice job here! I look forward to when this is enabled by default.

kpamnany commented:
@KristofferC: that kd-tree code would be a nice performance benchmark for ``test/perf/threads/`` if you can share it. PRs for performance tests would be very welcome.
